compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/HeaderProcessor.kt

```diff
@@ -149,44 +164,38 @@ internal class ProgressionLoopHeader(
                 //       // Loop body
                 //     } while (loopVar != last)
                 //   }
-                assert(loopVariable != null)
-                val booleanNotFun = context.irBuiltIns.booleanClass.functions.first { it.owner.name.asString() == "not" }
-                val newCondition = irCallOp(booleanNotFun, booleanNotFun.owner.returnType, irCall(context.irBuiltIns.eqeqSymbol).apply {
-                    putValueArgument(0, irGet(loopVariable!!))
-                    putValueArgument(1, irGet(last))
-                })
                 val newLoop = IrDoWhileLoopImpl(oldLoop.startOffset, oldLoop.endOffset, oldLoop.type, oldLoop.origin).apply {
                     label = oldLoop.label
-                    condition = newCondition
+                    condition = irNotEquals(irGet(loopVariable!!), irGet(last))
                     body = newBody
                 }
-                val notEmptyCheck = irIfThen(buildLoopCondition(builder), newLoop)
+                val notEmptyCheck = irIfThen(buildLoopCondition(this@with), newLoop)
```

Comment:
Why can you ignore the additional not empty condition here?

Reply:
Actually I should also use it here, but since it is only currently used in `until` progressions which go through the "cannot overflow" code path then we didn't expose this error. I'll updates this. Good catch!